pages/_app.tsx

```diff
@@ -1,10 +1,68 @@
 import CssBaseline from '@material-ui/core/CssBaseline'
-import { ThemeProvider } from '@material-ui/core/styles'
+import { ThemeProvider, createMuiTheme } from '@material-ui/core/styles'
 import App from 'next/app'
 import Head from 'next/head'
-import React from 'react'
+import React, { useState, createContext } from 'react'
+import { red, blue, deepPurple } from '@material-ui/core/colors'
+import { Color, PaletteType } from '@material-ui/core'
+import { NextComponentType, NextPageContext } from 'next'
 
-import theme from '../components/theme'
+interface AppContextStructure {
+	themeType: 'dark' | 'light'
+	setThemeType: React.Dispatch<React.SetStateAction<PaletteType>>
+	themePrimary: Color
+	setPrimary: React.Dispatch<React.SetStateAction<Color>>
+	themeSecondary: Color
+	setSecondary: React.Dispatch<React.SetStateAction<Color>>
+}
+
+interface Props {
+	component: NextComponentType<NextPageContext, any, {}>
+	pageProps: any
+}
+
+export const AppContext = createContext<Partial<AppContextStructure>>({})
+
+const AppWithHooks = (props: Props): JSX.Element => {
+	const [themeType, setThemeType] = useState<'dark' | 'light'>('dark')
+	const [primary, setPrimary] = useState<Color>(blue)
+	const [secondary, setSecondary] = useState<Color>(deepPurple)
+	const theme = createMuiTheme({
```

Comment:
I'd prefer theme be it's own module. `export default createMuiTheme()` somewhere and import it here. If you need to pass stuff into the theme then a factory function may be the best bet (`export default (stuff: StuffInterface) => createMuiTheme()`)

Reply:
theme must be part of a component to be reactive. It cannot be part of a spearate file.

Reply:
Why wouldn't a factory function work? Even if it was just building an object implementing the interface `createMuiTheme() expects, it'd still get the theme into a sensible location.

Reply:
the way components work is that anything outside the component is not reactive. the MUI theme must be constructed inside of the component or else it will not react to theme changes.

Reply:
Sure, the theme can be _constructed_ within the component, but I don't see why it must be _defined_ within it. e.g:

```jsx
// /theme.tsx
import { createMuiTheme, Theme } from '@material-ui/core/styles'
import { red, blue, deepPurple } from '@material-ui/core/colors'

export default function ThemeBuilder(isDarkMode: boolean): Theme {
	return createMuiTheme({
		palette: {
			primary: blue,
			secondary: deepPurple,
			type: isDarkMode ? 'dark' : 'light',
			error: {
				main: red.A400,
			},
		},
	})
}
```

```tsx
// /pages/_app.tsx
import ThemeBuilder from '../theme'

//...
	const [themeType, setThemeType] = useState<'dark' | 'light'>('dark')
	const theme = ThemeBuilder(themeType === 'dark')

	return (
		<>
			<Head>
				{/* https://err.sh/next.js/no-document-title */}
				<title>Impact</title>
			</Head>
			<ThemeProvider theme={theme}>
				{/* CssBaseline kickstart an elegant, consistent, and simple baseline to build upon. */}
				<CssBaseline />
				<AppContext.Provider
					value={{
						setThemeType,
						themeType,
					}}
				>
					<props.component {...props.pageProps} />
				</AppContext.Provider>
			</ThemeProvider>
		</>
	)
```